src/main/java/com/google/api/codegen/transformer/TestCaseTransformer.java

```diff
@@ -222,4 +223,86 @@ private InitCodeContext createResponseInitCodeContext(
     }
     return additionalSubTrees;
   }
+
+  public TestCaseView createSmokeTestCaseView(MethodTransformerContext context) {
+    ClientMethodType methodType = ClientMethodType.FlattenedMethod;
+    if (context.getMethodConfig().isPageStreaming()) {
+      methodType = ClientMethodType.PagedFlattenedMethod;
+    }
+
+    return createTestCaseView(
+        context, new SymbolTable(), createSmokeTestInitContext(context), methodType);
+  }
+
+  public boolean requireProjectId(InitCodeView initCodeView, SurfaceNamer namer) {
+    for (FieldSettingView settingsView : initCodeView.fieldSettings()) {
+      InitCodeLineView line = settingsView.initCodeLine();
+      if (line.lineType() == InitCodeLineType.SimpleInitLine) {
+        SimpleInitCodeLineView simpleLine = (SimpleInitCodeLineView) line;
+        String projectVarName =
+            namer.localVarName(Name.from(InitFieldConfig.PROJECT_ID_VARIABLE_NAME));
+        if (simpleLine.initValue() instanceof ResourceNameInitValueView) {
+          ResourceNameInitValueView initValue = (ResourceNameInitValueView) simpleLine.initValue();
+          return initValue.formatArgs().contains(projectVarName);
+        } else if (simpleLine.initValue() instanceof ResourceNameOneofInitValueView) {
+          ResourceNameOneofInitValueView initValue =
+              (ResourceNameOneofInitValueView) simpleLine.initValue();
+          ResourceNameInitValueView subValue = initValue.specificResourceNameView();
+          return subValue.formatArgs().contains(projectVarName);
+        } else if (simpleLine.initValue() instanceof SimpleInitValueView) {
+          SimpleInitValueView initValue = (SimpleInitValueView) simpleLine.initValue();
+          return initValue.initialValue().equals(projectVarName);
+        } else if (simpleLine.initValue() instanceof FormattedInitValueView) {
+          FormattedInitValueView initValue = (FormattedInitValueView) simpleLine.initValue();
+          return initValue.formatArgs().contains(projectVarName);
+        }
+      }
+    }
+    return false;
+  }
+
+  public InitCodeContext createSmokeTestInitContext(MethodTransformerContext context) {
+    SmokeTestConfig testConfig = context.getInterfaceConfig().getSmokeTestConfig();
+    InitCodeContext.InitCodeOutputType outputType;
+    ImmutableMap<String, FieldConfig> fieldConfigMap;
+    if (context.getMethodConfig().isFlattening()) {
+      outputType = InitCodeContext.InitCodeOutputType.FieldList;
+      fieldConfigMap =
+          FieldConfig.toFieldConfigMap(
+              context.getFlatteningConfig().getFlattenedFieldConfigs().values());
+    } else {
+      outputType = InitCodeContext.InitCodeOutputType.SingleObject;
+      fieldConfigMap = null;
+    }
+
+    // Store project ID variable name into the symbol table since it is used by the execute method
+    // as a parameter. For more information please see smoke_test.snip.
```

Comment:
Remove the comment 'For more information please see smoke_test.snip.' since now this method is cross-language.